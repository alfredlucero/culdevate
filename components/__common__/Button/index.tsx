import React from "react";
import styled, { css } from "../../../@types/styled-components/index";
import { culdevateThemes } from "../../defaultTheme";

interface StyledButtonProps {
  onClick: (event: React.MouseEvent) => void;
  disabled: boolean;
  kind: ButtonKind;
  type: ButtonType;
  size: ButtonSize;
}
const StyledButton = styled.button<StyledButtonProps>`
  border: none;
  border-radius: 0.25rem;
  cursor: pointer;
  font-weight: 600;

  ${({ size }) =>
    size === "small" &&
    css`
      padding: 1rem;
      font-size: 1rem;
    `}

  ${({ size }) =>
    size === "medium" &&
    css`
      padding: 1.5rem;
      font-size: 1.5rem;
    `}

  ${({ size }) =>
    size === "large" &&
    css`
      padding: 2rem;
      font-size: 2rem;
    `} 

  ${({
    kind,
    disabled,
    theme: {
      semanticColors: { btnDisabledBgColor, btnPrimaryBgColor, btnPrimaryBgHoverColor, btnPrimaryColor },
    },
  }) =>
    kind === "primary" &&
    css`
      background-color: ${disabled ? btnDisabledBgColor : btnPrimaryBgColor};
      color: ${btnPrimaryColor};

      &:hover {
        background-color: ${disabled ? btnDisabledBgColor : btnPrimaryBgHoverColor};
      }
    `}

  ${({
    kind,
    disabled,
    theme: {
      semanticColors: { btnDangerBgColor, btnDangerColor, btnDisabledBgColor, btnDangerBgHoverColor },
    },
  }) =>
    kind === "danger" &&
    css`
      background-color: ${disabled ? btnDisabledBgColor : btnDangerBgColor};
      color: ${btnDangerColor};

      &:hover {
        background-color: ${disabled ? btnDisabledBgColor : btnDangerBgHoverColor};
      }
    `}

  ${({
    kind,
    disabled,
    theme: {
      semanticColors: {
        btnSecondaryBgColor,
        btnSecondaryBgHoverColor,
        btnSecondaryColor,
        btnSecondaryBorderColor,
        btnSecondaryDisabledColor,
        btnSecondaryDisabledBgColor,
        btnSecondaryBorderDisabledColor,
      },
    },
  }) =>
    kind === "secondary" &&
    css`
      border: ${disabled ? btnSecondaryBorderDisabledColor : btnSecondaryBorderColor} 0.2rem solid;
      background-color: ${btnSecondaryBgColor};
      color: ${disabled ? btnSecondaryDisabledColor : btnSecondaryColor};

      &:hover {
        background-color: ${disabled ? btnSecondaryDisabledBgColor : btnSecondaryBgHoverColor};
      }
    `}

    ${({
      kind,
      disabled,
      theme: {
        semanticColors: {
          btnSecondaryDangerBgColor,
          btnSecondaryDangerColor,
          btnSecondaryDangerBorderColor,
          btnSecondaryDisabledColor,
          btnSecondaryDisabledBgColor,
          btnSecondaryDangerBgHoverColor,
          btnSecondaryBorderDisabledColor,
        },
      },
    }) =>
      kind === "secondaryDanger" &&
      css`
        border: ${disabled ? btnSecondaryBorderDisabledColor : btnSecondaryDangerBorderColor} 0.2rem solid;
        background-color: ${btnSecondaryDangerBgColor};
        color: ${disabled ? btnSecondaryDisabledColor : btnSecondaryDangerColor};

        &:hover {
          background-color: ${disabled ? btnSecondaryDisabledBgColor : btnSecondaryDangerBgHoverColor};
        }
      `}
`;
StyledButton.defaultProps = {
  theme: culdevateThemes.light,
};

export type ButtonKind = "primary" | "danger" | "secondary" | "secondaryDanger";
export type ButtonType = "button" | "submit" | "reset";
export type ButtonSize = "small" | "medium" | "large";

interface ButtonProps {
  onClick: (event: React.MouseEvent) => void;
  children: React.ReactNode;
  kind: ButtonKind;
  type: ButtonType;
  size: ButtonSize;
  disabled?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  onClick,
  children,
  kind,
  type,
  size,
  disabled = false,
  ...passThroughProps
}) => {
  return (
    <StyledButton onClick={onClick} kind={kind} type={type} size={size} disabled={disabled} {...passThroughProps}>
      {children}
    </StyledButton>
  );
};

export default Button;
