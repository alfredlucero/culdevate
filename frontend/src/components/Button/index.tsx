import React from "react";
import "./index.css";

interface ButtonProps {
  variant: ButtonVariant;
  children: React.ReactNode;
  onClick: (e: React.MouseEvent) => void;
}

type ButtonVariant = "primary" | "secondary" | "danger";

const Button: React.FC<ButtonProps> = ({ variant, children, onClick, ...passThroughProps }) => {
  return (
    <button className="btn btn-primary btn-primary:hover" {...passThroughProps} onClick={onClick}>
      {children}
    </button>
  );
};

export default Button;
