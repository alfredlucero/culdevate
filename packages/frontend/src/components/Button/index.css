.btn {
  @apply font-bold py-2 px-4 rounded;
}

.btn-disabled {
  @apply opacity-50 cursor-not-allowed;
}

.btn-primary {
  @apply bg-blue-500 text-white;
}

.btn-primary:hover {
  @apply bg-blue-700;
}

.btn-secondary {
  @apply bg-transparent text-blue-700 border border-blue-500;
}

.btn-secondary:hover {
  @apply bg-blue-500 text-white border-transparent;
}
