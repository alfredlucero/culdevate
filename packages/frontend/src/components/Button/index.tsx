import React from "react";
import cn from "classnames";
import { CommonProps } from "../commonProps";

interface ButtonProps extends CommonProps {
  variant: ButtonVariant;
  type: ButtonType;
  children: React.ReactNode;
  onClick: (e: React.MouseEvent) => void;
  disabled?: boolean;
}

type ButtonType = "submit" | "button" | "reset";
type ButtonVariant = "primary" | "secondary" | "danger";

const buttonBaseClasses = ["font-bold", "py-2", "px-4", "rounded"];
const buttonDisabledClasses = ["opacity-50", "cursor-not-allowed"];
const buttonPrimaryClasses = [...buttonBaseClasses, "bg-teal-500", "text-white", "hover:bg-teal-700"];
const buttonSecondaryClasses = [
  "bg-transparent",
  "text-teal-700",
  "border",
  "border-teal-500",
  "hover:bg-teal-500",
  "hover:text-white",
  "hover:border-transparent",
];
const buttonDangerClasses = [...buttonBaseClasses, "bg-red-500", "text-white", "hover:bg-red-700"];

const Button: React.FC<ButtonProps> = ({
  variant,
  type,
  children,
  onClick,
  disabled = false,
  className = "",
  testId = "",
  ...passThroughProps
}) => {
  return (
    <button
      className={`${cn(
        buttonBaseClasses,
        ...(variant === "primary" ? buttonPrimaryClasses : []),
        ...(variant === "secondary" ? buttonSecondaryClasses : []),
        ...(variant === "danger" ? buttonDangerClasses : []),
        disabled ? buttonDisabledClasses : [],
      )} ${className}`}
      onClick={onClick}
      disabled={disabled}
      type={type}
      {...(testId !== "" ? { "data-testid": testId } : {})}
      {...passThroughProps}
    >
      {children}
    </button>
  );
};

export default Button;
