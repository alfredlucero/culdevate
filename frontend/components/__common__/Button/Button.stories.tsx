import React from "react";
import { storiesOf } from "@storybook/react";
import StoryFrame from "../StoryFrame";
import Button from "./index";

storiesOf("common/Button", module)
  .add("Primary Button Small", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="button" kind="primary" size="small">
        Primary
      </Button>
    </StoryFrame>
  ))
  .add("Primary Submit Medium", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="primary" size="medium">
        Primary
      </Button>
    </StoryFrame>
  ))
  .add("Primary Submit Medium Disabled", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="primary" size="medium" disabled={true}>
        Primary
      </Button>
    </StoryFrame>
  ))
  .add("Primary Reset Large", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="reset" kind="primary" size="large">
        Primary
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Button Small", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="button" kind="secondary" size="small">
        Secondary
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Submit Medium", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="secondary" size="medium">
        Secondary
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Submit Medium Disabled", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="secondary" size="medium" disabled={true}>
        Secondary
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Reset Large", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="reset" kind="secondary" size="large">
        Secondary
      </Button>
    </StoryFrame>
  ))
  .add("Danger Button Small", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="button" kind="danger" size="small">
        Danger
      </Button>
    </StoryFrame>
  ))
  .add("Danger Submit Medium", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="danger" size="medium">
        Danger
      </Button>
    </StoryFrame>
  ))
  .add("Danger Submit Medium Disabled", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="danger" size="medium" disabled={true}>
        Danger
      </Button>
    </StoryFrame>
  ))
  .add("Danger Reset Large", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="reset" kind="danger" size="large">
        Danger
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Danger Button Small", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="button" kind="secondaryDanger" size="small">
        Secondary Danger
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Danger Submit Medium", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="secondaryDanger" size="medium">
        Secondary Danger
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Danger Submit Medium Disabled", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="submit" kind="secondaryDanger" size="medium" disabled={true}>
        Secondary Danger
      </Button>
    </StoryFrame>
  ))
  .add("Secondary Danger Reset Large", () => (
    <StoryFrame>
      <Button onClick={() => {}} type="reset" kind="secondaryDanger" size="large">
        Secondary Danger
      </Button>
    </StoryFrame>
  ));
