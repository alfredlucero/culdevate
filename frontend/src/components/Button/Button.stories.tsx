import React from "react";
import { storiesOf } from "@storybook/react";
import Button from "./index";

storiesOf("Common/Button", module).add("Default", () => (
  <Button variant="primary" onClick={() => {}}>
    Primary
  </Button>
));
