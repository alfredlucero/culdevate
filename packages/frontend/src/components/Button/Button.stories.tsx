import React from "react";
import { storiesOf } from "@storybook/react";
import Button from "./index";

storiesOf("Common/Button", module)
  .add("Primary", () => (
    <Button type="button" variant="primary" onClick={() => {}}>
      Primary
    </Button>
  ))
  .add("Primary Disabled", () => (
    <Button type="submit" variant="primary" onClick={() => {}} disabled={true}>
      Primary Disabled
    </Button>
  ))
  .add("Secondary", () => (
    <Button type="reset" variant="secondary" onClick={() => {}}>
      Secondary
    </Button>
  ))
  .add("Secondary Disabled", () => (
    <Button type="button" variant="secondary" onClick={() => {}} disabled={true}>
      Secondary Disabled
    </Button>
  ))
  .add("Danger", () => (
    <Button type="reset" variant="danger" onClick={() => {}}>
      Danger
    </Button>
  ))
  .add("Danger Disabled", () => (
    <Button type="button" variant="danger" onClick={() => {}} disabled={true}>
      Danger Disabled
    </Button>
  ));
