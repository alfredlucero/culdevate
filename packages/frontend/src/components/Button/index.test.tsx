import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Button from "./index";

describe("<Button />", () => {
  const buttonTestId = "buttonTestId";

  test("should render primary variant without error", () => {
    const { container } = render(
      <Button variant="primary" type="button" testId={buttonTestId} className="extra-button-class" onClick={() => {}}>
        Primary Button
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });

  test("should render secondary variant without error", () => {
    const { container } = render(
      <Button variant="secondary" type="submit" testId={buttonTestId} className="extra-button-class" onClick={() => {}}>
        Secondary Button
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });

  test("should render danger variant without error", () => {
    const { container } = render(
      <Button variant="danger" type="reset" testId={buttonTestId} className="extra-button-class" onClick={() => {}}>
        Danger Button
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });

  test("should render loading state without error", () => {
    const { container } = render(
      <Button
        variant="primary"
        type="button"
        loading={true}
        testId={buttonTestId}
        className="extra-button-class"
        onClick={() => {}}
      >
        Primary Loading Button
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });

  test("should render disabled state when disabled", () => {
    const { getByTestId } = render(
      <Button variant="primary" type="button" testId={buttonTestId} onClick={() => {}} disabled={true}>
        Disabled Button
      </Button>,
    );

    expect(getByTestId(buttonTestId)).toBeDisabled();
  });

  test("should trigger onClick when click event is fired", () => {
    const onClickMock = jest.fn();
    const { getByTestId } = render(
      <Button variant="primary" type="button" testId={buttonTestId} onClick={onClickMock}>
        Primary Button
      </Button>,
    );

    fireEvent.click(getByTestId(buttonTestId), { button: 0 });

    expect(onClickMock).toHaveBeenCalled();
  });
});
