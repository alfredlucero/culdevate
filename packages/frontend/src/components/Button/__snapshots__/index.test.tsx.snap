// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render danger variant without error 1`] = `
<div>
  <button
    class="font-bold py-2 px-4 rounded font-bold py-2 px-4 rounded bg-red-500 text-white hover:bg-red-700 extra-button-class"
    data-testid="buttonTestId"
    type="reset"
  >
    Danger Button
  </button>
</div>
`;

exports[`<Button /> should render primary variant without error 1`] = `
<div>
  <button
    class="font-bold py-2 px-4 rounded font-bold py-2 px-4 rounded bg-teal-500 text-white hover:bg-teal-700 extra-button-class"
    data-testid="buttonTestId"
    type="button"
  >
    Primary Button
  </button>
</div>
`;

exports[`<Button /> should render secondary variant without error 1`] = `
<div>
  <button
    class="font-bold py-2 px-4 rounded bg-transparent text-teal-700 border border-teal-500 hover:bg-teal-500 hover:text-white hover:border-transparent extra-button-class"
    data-testid="buttonTestId"
    type="submit"
  >
    Secondary Button
  </button>
</div>
`;
